src/peer_digest.cc

```diff
@@ -134,40 +129,20 @@ peerDigestCreate(CachePeer * p)
 
     // TODO: make CachePeer member a CbcPointer
     p->digest = cbdataReference(pd);
-
-    // lock a reference to pd again to prevent the PeerDigest
-    // disappearing during peerDigestDestroy() when
-    // cbdataReferenceValidDone is called.
-    // TODO test if it can be moved into peerDigestDestroy() or
-    //      if things can break earlier (eg CachePeer death).
-    (void)cbdataReference(pd);
 }
 
 /* call Clean and free/unlock everything */
 static void
 peerDigestDestroy(PeerDigest * pd)
 {
-    void *p;
     assert(pd);
-    void * peerTmp = pd->peer;
-
-    /*
-     * DPW 2007-04-12
-     * We locked the peer in PeerDigest constructor, this is
-     * where we unlock it.
-     */
-    if (cbdataReferenceValidDone(peerTmp, &p)) {
-        // we locked the p->digest in peerDigestCreate()
-        // this is where we unlock that
-        cbdataReferenceDone(static_cast<CachePeer *>(p)->digest);
```

Comment:
> > AFAICT, the above commit did not create the bug that this PR description is blaming for the leak. In that commit, peerDigestCreate() never locked the digest and peerDigestDestroy() never unlocked it. That commit behavior sounds correct and "balanced" to me, but perhaps I missed something in that (overall unreadable) code.

> Here is the related conditional change from https://github.com/measurement-factory/squid/commit/e13ee7ad48fb523b6cba51056584044ea7ed85f4:

```C++
void
peerDigestDestroy(PeerDigest *pd)
{
    assert(pd);
    assert(cbdataValid(pd));

    /* inform peer (if any) that we are gone */
    if (cbdataValid(pd->peer))
       peerNoteDigestGone(pd->peer);
    cbdataUnlock(pd->peer); /* must unlock, valid or not */
    pd->peer = NULL;

    peerDigestClean(pd);
    cbdataFree(pd);
}
```

In the above e13ee7a code, the effective dereference[[^1] of the `pd->peer` pointer is conditional on the pointer validity and the following unlock of that peer pointer is unconditional. Both look correct to me.

[^1] peerNoteDigestGone(pd->peer) is effectively a pd->peer->noteDigestGone() call; the function dereferences its peer parameter without checking, as if it was a "this" pointer in a class method.

Reply:
One possible source of confusion here is that the current official code _looks_ correct, but there is a critical difference:

```C++
// on the surface, the current official code looks like this:
if (valid(peer))
    doSomethingWith(peer->digest);
```

The above is correct, in principle, but the buggy official code this change request is attached to should not be interpreted this way! It should be interpreted this way instead:

```C++
// in reality, the current official code does this:
if (valid(peer))
    unlock(digest);
```

... which is a bug this PR is fixing. In other words, the official code locks the digest object and should unlock that digest object regardless of CachePeer object state. The apparent connection between CachePeer and digest objects falsely implied by the official code does not actually exist as far as unlocking of the digest object is concerned.

Reply:
> ) unlock worked together with another digest unlock in peerDestroy() 

Did you mean 'in peerDigestCheck()'? in peerDestroy(), `peer` must be still valid...

> // else peerDestroy() already unlocked our digest

Again, if you meant `peerDigestCheck()` - there are a couple of places in that revision, where the digest is unlocked, but it is not clear from that context that the 'invalid peer' case is always  covered... But I did not study that code thoroughly and may miss something, of course.

Reply:
> > ) unlock worked together with another digest unlock in peerDestroy()
> 
> Did you mean 'in peerDigestCheck()'? in peerDestroy(), `peer` must be still valid...

No, I meant peerDestroy(). Yes, the peer in peerDestroy() should be valid. peerDestroy() is what invalidates the peer. In e13ee7a, peerNoteDigestGone() and peerDestroy() work in tandem: Only one of them can succeed in unlocking the digest. Each unlock the digest when peer is still valid. If we reach peerDestroy() first, peerDestroy() unlocks. If we reach peerNoteDigestGone() first (with a still valid peer), peerNoteDigestGone() unlocks.

BTW, the peer is always valid in peerNoteDigestGone() as well, but that function code does not know that. Here is the only peerNoteDigestGone() caller in e13ee7a:

```C++
/* inform peer (if any) that we are gone */
if (cbdataValid(pd->peer))
    peerNoteDigestGone(pd->peer);
```

> > // else peerDestroy() already unlocked our digest
> 
> Again, if you meant `peerDigestCheck()` - there are a couple of places in that revision, where the digest is unlocked, but it is not clear from that context that the 'invalid peer' case is always covered...

I meant peerDestroy(). The case of an invalid peer is always covered in e13ee7a because the only way to invalidate a peer is to call peerDestroy()[^1], and peerDestroy() unlocks the digest (if it has not been unlocked already). We do not need to study any other digest unlock calls to reach this particular conclusion AFAICT.

[^1]: `struct _peer` does not even have a destructor. The peerDestroy() is, effectively, that destructor, and it (naturally) calls cbdataFree() after cleaning up the structure. cbdataFree() invalidates the peer.